auth/src/main/java/com/firebase/ui/auth/data/remote/FacebookSignInHandler.java

```diff
@@ -0,0 +1,167 @@
+
+    private final CallbackManager mCallbackManager = CallbackManager.Factory.create();
+
+    public FacebookSignInHandler(Application application) {
+        super(application);
+    }
+
+    private static IdpResponse createIdpResponse(
+            LoginResult result, @Nullable String email, String name, Uri photoUri) {
+        return new IdpResponse.Builder(
+                new User.Builder(FacebookAuthProvider.PROVIDER_ID, email)
+                        .setName(name)
+                        .setPhotoUri(photoUri)
+                        .build())
+                .setToken(result.getAccessToken().getToken())
+                .build();
+    }
+
+    @Override
+    protected void onCreate() {
+        initPermissionList();
+        LoginManager.getInstance().registerCallback(mCallbackManager, this);
+    }
+
+    private void initPermissionList() {
+        List<String> scopes = getArguments().config.getParams()
+                .getStringArrayList(ExtraConstants.EXTRA_FACEBOOK_PERMISSIONS);
+        if (scopes == null) {
+            scopes = new ArrayList<>();
+        }
+
+        List<String> permissionsList = new ArrayList<>(scopes);
+
+        // Ensure we have email and public_profile scopes
+        if (!permissionsList.contains(EMAIL)) {
+            permissionsList.add(EMAIL);
+        }
+        if (!permissionsList.contains(PUBLIC_PROFILE)) {
+            permissionsList.add(PUBLIC_PROFILE);
+        }
+
+        mPermissions = permissionsList;
+    }
+
+    public List<String> getPermissions() {
+        return mPermissions;
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
+        mCallbackManager.onActivityResult(requestCode, resultCode, data);
+    }
+
+    @Override
+    public void onSuccess(final LoginResult result) {
+        GraphRequest request = GraphRequest.newMeRequest(result.getAccessToken(),
+                new ProfileRequest(result));
+
+        Bundle parameters = new Bundle();
+        parameters.putString("fields", "id,name,email,picture");
+        request.setParameters(parameters);
+        request.executeAsync();
+    }
+
+    @Override
+    public void onCancel() {
+        onError(new FacebookException());
+    }
+
+    @Override
+    public void onError(FacebookException e) {
+        setResult(IdpResponse.fromError(e));
+    }
+
+    @Override
+    protected void onCleared() {
+        super.onCleared();
+        LoginManager.getInstance().unregisterCallback(mCallbackManager);
+    }
+
+    private class ProfileRequest implements GraphRequest.GraphJSONObjectCallback {
+        private final LoginResult mResult;
+
+        public ProfileRequest(LoginResult result) {
+            mResult = result;
+        }
+
+        @Override
+        public void onCompleted(JSONObject object, GraphResponse response) {
+            FacebookRequestError error = response.getError();
+            if (error != null || object == null) {
```

Comment:
This `if` statement is a bit over packed.  It has conditions on `error` inside and outside.

Suggest separate if-return-error statements for the different error conditions.